# Installation

Alternative source for installation instructions: [Recentrifuge Wiki Installation page](https://github.com/khyox/recentrifuge/wiki/Installation)

## Overview
1.  **Basic requirements**: Python 3.6 is required. No modules beyond Python Standard Library ones are used by Recentrifuge, with the exception of _biopython_ for LMAT support and, optionally, _pandas_ for exporting results to Excel.
2. **Getting the code**: just clone the repository.
3. **Getting the databases**: in the cloning dir, execute `retaxdump.py`. It will download and unzip the required local databases from NCBI servers under the subdirectory `taxdump`. For LMAT plasmids support please read [this](#support-for-lmat-plasmids-classification).
4. At this point, **Recentrifuge is ready** to analyze your samples:
  * if you have results from Centrifuge, please see [[running Recentrifuge for Centrifuge|Running-recentrifuge-for-Centrifuge]],
  * in case you have LMAT outputs, your choice is [[running Recentrifuge for LMAT|Running-recentrifuge-for-LMAT]],
  * if you have CLARK(S) results, please enter [[running Recentrifuge for CLARK|Running-recentrifuge-for-CLARK]].
5. **Interactive visualization** of Recentrifuge results: Just open the HTML file generated by Recentrifuge with any JavaScript-enabled browser. Firefox or Chrome are recommended.
6. **Numeric results** for downstream applications: Recentrifuge generates an Excel file with various sheets containing diverse statistics and detailed numeric results. 

## Details
### Python version
Python version under 3.6 is no supported as Recentrifuge uses new syntax features of Python 3.6, like syntax for variable annotations ([PEP 526](https://www.python.org/dev/peps/pep-0526/)) and formatted string literals ([PEP 498](https://www.python.org/dev/peps/pep-0498/)). The syntax for type annotations was introduced in Python 3.5 ([PEP 484](https://www.python.org/dev/peps/pep-0484/)) but it is with Python 3.6 when it has achieved maturity for variable annotations. Powerful tools for [static type analysis](https://en.wikipedia.org/wiki/Static_program_analysis) in Python have evolved along with these standards. The development of Recentrifuge includes checks with [pylint](https://www.pylint.org/) and [mypy](http://www.mypy-lang.org). A code whose aim is to perform robust comparative metagenomic analysis is a very good candidate for robust coding.

### Support for LMAT plasmids classification
One of the most interesting but still quite unknown features of the LMAT software is its ability to properly classify over 4000 plasmids. These plasmids are assigned a taxonomical id (`taxid`) beyond the NCBI system. Recentrifuge offers support for this extended classification, but requires the LMAT provided file `plasmid.names.txt` located in the same directory as the NCBI nodes information files. This location is controlled by the flag `-n/--nodespath`.

If Recentrifuge finds the `plasmid.names.txt` file, it will parse the plasmids taxid and name using ad hoc regular expressions in order to present the user with a meaningful name for the very diverse plasmids. Recentrifuge is also doing a check to assure that every plasmid in the file is compatible with the NCBI taxonomy used, so that only those passing are added. Further details [[here|Running-recentrifuge-for-LMAT#more-about-plasmids]].
